### ISSUE TITLE
Speed and memory requirements optimizing. 

### ISSUE THREAD

**Issue by robinvanleeuwen** [notadamking/RLTrader#53]
Hi, i am running new optimization against a dataset in which the sorting is done correctly (with regard to the AM/PM bug in sorting). And i noticed that when i run optimization with 
1 job, 1 trial, 1 episode and 1 eval. It takes about 13 minutes and 4G of memory to complete the cycle (Added some logging to the optimize.py)  :

DEBUG:__main__: input_data_file: data/coinbase_hourly.csv
DEBUG:__main__: Dropping 'Symbol' column from dataframe
DEBUG:__main__: Converting 'Date' column to type datetime
DEBUG:__main__: Sorting dataframe by 'Date' column
DEBUG:__main__: Adding indicators to dataframe
INFO:__main__: Reward strategy : sortino
INFO:__main__: Input data file : data/coinbase_hourly.csv
INFO:__main__: Parames db file : sqlite:///params.db
INFO:__main__: n_jobs          = 1
INFO:__main__: n_trials        = 1
INFO:__main__: n_test_episodes = 1
INFO:__main__: n_evaluations   = 1
INFO:__main__: Total number of records : 13144
INFO:__main__: Number of training recs : 13144
INFO:__main__: Num. of evaluation recs : 10515
...
DEBUG:__main__: Running 1 episodes
DEBUG:__main__: Started optimization : 2019-06-17 12-52-49
DEBUG:__main__: Finished optimization: 2019-06-17 13:06:26
DEBUG:__main__: Total trial time : 0:13:36.920658
INFO:optuna.study:Finished trial#2 resulted in value: 1886.0362548828125. Current best value is 0.0 with parameters: {'cliprang$
Number of finished trials:  1
Best trial:
Value:  0.0
Params:
    cliprange: 0.34253769458062744
    confidence_interval: 0.8379242988756598
    ent_coef: 3.63811777280501e-07
    forecast_len: 1.1511618552611296
    gamma: 0.9913080137473371
    lam: 0.9312156811763672
    learning_rate: 0.027396718735216347
    n_steps: 122.04674817782056
    noptepochs: 1.4269422834689458
DEBUG:__main__: Total run time (1 jobs, 1 trials, 1 evals, 1 episodes) : 0:13:37.238973

When i run the same script with n_trials = 3, i would excpect that it takes 3*13.37 minutes. However after about 1 hour, my computer froze with 8GB ram filled up, and 7.5G swapping to disk. (no other processes running). 

That leads me to two questions: 

1. When running multiple trials on modest hardware (4-core i7 , 8GB RAM , no GPU) would it be wise to loop through optimize() :

     for i in range(0..number_of_trials):
         optimize()

with n_trials in optimize() defined as 1 instead of: 

     study.optimize(optimize_agent, n_trials=n_trials, n_jobs=n_jobs)

Since the results are just added tot params.db

2. What would a decent set of hardware be to run say 100 trials. And how many trials would it take to return a reasonably valid set of hyperparameters?

**Comment by TalhaAsmal** [comment]
There are 2 parameters that you might be getting confused with: n_trials, and n_jobs

n_trials specifies how many different trials to test using optuna, so essentially how many 'buckets' to create for the different parameters.

n_jobs indicates how many concurrent trials to run.

In your case, have you tried setting n_jobs = 1 always, and increasing the number of trials? Arguably, the more trials you run, the better your chances of finding optimal parameters, at the cost of taking longer to test all combinations of trials.

If you're already running a single job at a time (n_jobs = 1), then the only thing I can think of is some combination of parameters is using way too much memory. If you're interested in finding that combination that causes your machine to hang, you can print out the parameters after they've been defined:

    model_params = optimize_ppo2(trial)
    print(f"[{datetime.now()}] Trial {trial_num} model_params: {model_params}")

Your data might also be too large to fit effectively in memory. For reference, I use hourly data from 2016-01-01 and it takes about 15 minutes for a single trial, and fits in under 16GB of RAM

**Comment by robinvanleeuwen** [comment]
Mmm, you are correct.  I re-ran the optimizer and indeed saw no memory increase when using 
n_trials = 1  or n_trials = 30. Memory usage stayed 1.2G

And I did see a memory increase from 1.2GB to 7.5GB+ and then started swapping until swap reached 7.5G after which the process terminated without message, when using n_trials = 30 and n_jobs = 3

So i guess when running my first test I did get n_trials and n_jobs mixed up and probably was doing 
n_jobs = 4 and n_trials = 1 by accident instead of the other way round. 

But the large increase in memory use 7.5G+ and 7.5G swap when using n_jobs = 3 instead of n_jobs = 1 (on a 4-core processor) , does sound like a memory leak to me somewhere....

I'm going to run 40 trials with n_jobs = 1 tonight on a correctly sorted dataset ( #28  ). Which should take +- 9 hours. See where i end up training with hyper-params collected from that. 

Does anyone have optimized hyper parameters with a large number of trials on the correctly sorted dataset already?

**Comment by robinvanleeuwen** [comment]
@TalhaAsmal 
No... I was too soon to say that it worked correctly in my post above.  It does not play well with memory.

When i use n_jobs = 1 and n_trials = 40. The first trial goes as expected. Memory usage +/- 1GB and 1 core 100%. When the first trial is finished and it starts with the second trial then memory shoots up to :

-
(venv) rvl@ub06:~/devel/python/bitcoin/latest$ free -h
              total        used        free      shared  buff/cache   available
Mem:           7,7G        7,0G        102M        443M        585M         37M
Swap:          7,9G        6,5G        1,4G
-

After which my system becomes unusable until the process gives up.... Complete logging below (cut out the optimizing steps: 


(venv) rvl@ub06:~/devel/python/bitcoin/latest$ time python3 optimize.py | tee log/optimize.log 
DEBUG:__main__: input_data_file: data/coinbase_hourly.csv
DEBUG:__main__: Dropping 'Symbol' column from dataframe
DEBUG:__main__: Converting 'Date' column to type datetime
DEBUG:__main__: Sorting dataframe by 'Date' column
DEBUG:__main__: Adding indicators to dataframe
INFO:__main__: Reward strategy : sortino
INFO:__main__: Input data file : data/coinbase_hourly.csv
INFO:__main__: Parames db file : sqlite:///params.db
INFO:__main__: n_jobs          = 1
INFO:__main__: n_trials        = 40
INFO:__main__: n_test_episodes = 1
INFO:__main__: n_evaluations   = 1
INFO:__main__: Total number of records : 13144
INFO:__main__: Number of training recs : 13144
INFO:__main__: Num. of evaluation recs : 10515
DEBUG:optuna.storages.rdb.storage:Ignoring IntegrityError('(sqlite3.IntegrityError) UNIQUE constraint failed: studies.study_name'). This happens due to a timing issue among threads/processes/nodes. Another one might have committed a record with the same key(s).
INFO:optuna.study:Using an existing study with name 'ppo2_sortino' instead of creating a new one.
DEBUG:__main__: Optimizing agent
DEBUG:__main__: Starting optimization (2019-06-17 23:26:56)
DEBUG:__main__: Model parameters after optimize_ppo2(trial):
DEBUG:__main__: {'n_steps': 28, 'gamma': 0.9877738142146887, 'learning_rate': 0.5684967217736766, 'ent_coef': 2.625032330868934e-05, 'cliprange': 0.3958977092620506, 'noptepochs': 3, 'lam': 0.9647339228817504}
DEBUG:__main__: Starting evaluation number: 0
DEBUG:__main__: model.learn(10515)
------------------------------------
| approxkl           | 0.59316915  |
| clipfrac           | 0.42857143  |
| explained_variance | -0.000483   |
| fps                | 8           |
| n_updates          | 1           |
| policy_entropy     | 2.2830305   |
| policy_loss        | 0.048654348 |
| serial_timesteps   | 28          |
| time_elapsed       | 2.86e-06    |
| total_timesteps    | 28          |
| value_loss         | 255847.45   |
------------------------------------
 ....
--------------------------------------
| approxkl           | 0.0           |
| clipfrac           | 0.0           |
| explained_variance | 1.19e-07      |
| fps                | 5             |
| n_updates          | 375           |
| policy_entropy     | 1.7366463e-18 |
| policy_loss        | 2.0435878e-07 |
| serial_timesteps   | 10500         |
| time_elapsed       | 719           |
| total_timesteps    | 10500         |
| value_loss         | 23053.713     |
--------------------------------------
DEBUG:__main__: Model has learned
DEBUG:__main__: Running 1 episodes
DEBUG:__main__: Reporting reward: -80569.3046875
DEBUG:__main__: Started optimization : 2019-06-17 23-26-56
DEBUG:__main__: Finished optimization: 2019-06-17 23:40:18
DEBUG:__main__: Total trial time : 0:13:21.700079
DEBUG:__main__: Reporting reward: -80569.3046875
INFO:optuna.study:Finished trial#8 resulted in value: -80569.3046875. Current best value is -80569.3046875 with parameters: {'cliprange': 0.3958977092620506, 'confidence_interval': 0.8743439011379142, 'ent_coef': 2.625032330868934e-05, 'forecast_len': 12.697117873217922, 'gamma': 0.9877738142146887, 'lam': 0.9647339228817504, 'learning_rate': 0.5684967217736766, 'n_steps': 28.95580524603379, 'noptepochs': 3.936114575256249}.
DEBUG:__main__: Optimizing agent
DEBUG:__main__: Starting optimization (2019-06-17 23:40:18)
DEBUG:__main__: Model parameters after optimize_ppo2(trial):
DEBUG:__main__: {'n_steps': 1329, 'gamma': 0.9032099099895787, 'learning_rate': 1.3400011185980377e-05, 'ent_coef': 7.822982599712703e-07, 'cliprange': 0.32031703604289846, 'noptepochs': 10, 'lam': 0.8264005032973986}
DEBUG:__main__: Starting evaluation number: 0
DEBUG:__main__: model.learn(10515)
         <----------- PROCESS HALTS HERE AFTER CONSUMING ALL MEMORY
real	20m17,326s
user	19m54,023s
sys	1m59,987s
(venv) rvl@ub06:~/devel/python/bitcoin/latest$
